Windows offline activation with ACPI SLIC table

victortoso opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
It allows setting the SLIC ACPI table which is useful for offline Windows activation.

The user needs to have the data following [Microsoft specification](https://learn.microsoft.com/en-us/previous-versions/windows/hardware/design/dn653305(v=vs.85)). This PR does not provide a way for users to generate this binary data.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

This is a working PR with a proposal to provide this feature. We can test this without Windows VMs as the ACPI SLIC table can be verified in a Linux Guest.

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Windows offline activation with ACPI SLIC table
```

victortoso commented:
```
 ./hack/verify-generate.sh
./hack/check-for-binaries.sh
< tests/testdata/firmware-acpi-slic-table.bin:1
Binary files are present in git repostory.
make: *** [Makefile:56: generate-verify] Error 1 
```

This is a 36 bytes file... it should be fine for test purposes but I can move it to const array if binary files are no-no.

alicefr commented:
/cc

alicefr commented on tests/vmi_configuration_test.go:
```diff
@@ -619,6 +623,59 @@ var _ = Describe("[sig-compute]Configurations", decorators.SigCompute, func() {
 
+		Context("with ACPI SLIC table", func() {
+			It("Should configure guest APCI SLIC with Secret file", func() {
+				const volumeSlicSecretName = "volume-slic-secret"
+				const secretWithSlicName = "secret-with-slic-data"
+				const hexData = "534c49432400000001494352415348204d450000000000008804000071656d7500000000"
```
super nit: you can use a single const block

victortoso replied:
Hi Alice, thanks for the quick review. I'll make it a const block in the next iteration. I'll probably move the binary data here as well and use that to compare (instead of `hexData`).

alicefr commented:
@victortoso the code looks already great!
/lgtm

vasiliy-ul commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -427,6 +427,15 @@ type Firmware struct {
+}
+
+type ACPI struct {
+	// SliceNameRef should match the volume name of a secret object. The data in the secret should
```
```suggestion
	// SlicNameRef should match the volume name of a secret object. The data in the secret should
```

victortoso replied:
Hi @vasiliy-ul , thanks for the quick review.

I thought about referencing the Secret in the `SlicNameRef` as well but I decided to do the easier path first.

I don't have a strong opinion in this so I can change it. Let's just double check if others agree. @alicefr, @xpivarc  what do you think?

alicefr replied:
IMO, pointing to the volume name is fine. If for any reasons we decide to support another type of volume for this, then this mechanism still remains true. Also, kubernetes references the secrets by volume name inside the pod spec. So, I don't think it is too unusual for the user to reference the volume name instead of the secret name.

vasiliy-ul replied:
Just to note: changing later the reference from volume to secret will be a breaking API change. I am not pushing with this, but IMHO it is better to come up with a long-term decision when the API is introduced. Altering it after the release may become problematic.

victortoso replied:
I agree that deciding now is better. For the moment, I'm keeping the current code behavior to reference the volume name.

vasiliy-ul reviewed (commented):
@victortoso, the code looks good overall. Just a thought from my side: as I see now there is a need to manually setup the volume with volume source pointing to the proper secret. Wouldn't it be more convenient for the end-user if we had just `SlicNameRef` in the VMI spec referencing the **secret** (not to the volume name), and the volume could be rendered automatically if `vmi.Spec.Domain.Firmware.ACPI.SlicNameRef` is set?

victortoso commented:
- Rebased
- Fixed the nits, thanks Alice and Vasiliy for the reviews.
- Removed the embed 36 bytes binary to workaround the [check-for-binaries.sh](https://github.com/kubevirt/kubevirt/pull/10774#issuecomment-1823595630). The slic table is now a []byte in the test.

victortoso commented:
/retest

vasiliy-ul commented:
/lgtm

vasiliy-ul commented:
> Removed the embed 36 bytes binary to workaround the https://github.com/kubevirt/kubevirt/pull/10774#issuecomment-1823595630.

BTW, I guess you could have used base64 to store the table in a text file (and silence the checker) and use embed. Then decode it in the test... Just a thought.

victortoso commented:
> Still leaving it up to @victortoso to unhold :)

Thanks @vasiliy-ul and @alicefr for the quick reviews. Let's unhold it then. I'm not sure I can do it, but does not hurt to try:

/unhold

> BTW, I guess you could have used base64 to store the table in a text file (and silence the checker) and use embed. Then decode it in the test... Just a thought.

That's neat. I'll keep it in mind for the next time.

victortoso commented:
/retest

victortoso commented:
/test pull-kubevirt-check-tests-for-flakes

victortoso commented:
```
19:49:20: 19:49:20: 2023/11/30 19:49:20 Image pull was unsuccessful: reading image "quay.io/kubevirt/alpine-ext-kernel-boot-demo@sha256:a2ddb2f568bf3814e594a14bc793d5a655a61d5983f3561d60d02afa7bbc56b4": GET https://quay.io/v2/: unsupported status code 502; body:
```

/test pull-kubevirt-check-tests-for-flakes

victortoso commented:
/retest

xpivarc commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -427,6 +427,15 @@ type Firmware struct {
+	// SlicNameRef should match the volume name of a secret object. The data in the secret should
+	// be a binary blob that follows the ACPI SLIC standard, see:
+	// https://learn.microsoft.com/en-us/previous-versions/windows/hardware/design/dn653305(v=vs.85)
+	SlicNameRef string `json:"slicNameRef,omitempty"`
```
Do we need to restrict this to x86? @alicefr @victortoso

victortoso replied:
I don't think so. I'm double checking and I'll let you know.

victortoso replied:
Hi @xpivarc, there is no explicit restriction. It should work form Windows on arm, albeit this was not tested.
